File: java/src/test/java/shapes/ModelsTest.java
package shapes;

import static org.junit.jupiter.api.Assertions.assertTrue;
import static shapes.fixtures.Helper.setupCircle;
import static shapes.fixtures.Helper.setupRectangle;
import static shapes.fixtures.Helper.setupSquare;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import shapes.fixtures.Helper;

@DisplayName("Models Tests")
class ModelsTest {
    private static final double DELTA = 0.001;

    @Nested
    @Tag("Circle")
    @DisplayName("Circle Tests")
    class CircleTest {
        @Test
        @Tag("Area")
        @DisplayName("Test Circle Area")
        void testCircleArea() {
            var circle = setupCircle();
            double result = circle.area();
            double expected = 754.767;
            assertTrue(Math.abs(result - expected) < DELTA);
        }

        @Test
        @Tag("Perimeter")
        @DisplayName("Test Circle Circumference")
        void testCircleCircumference() {
            var circle = Helper.setupCircle();
            double result = circle.perimeter();
            double expected = 97.389;
            assertTrue(Math.abs(result - expected) < DELTA);
        }
    }

    @Nested
    @Tag("Rectangle")
    @DisplayName("Rectangle Tests")
    class RectangleTest {
        @Test
        @Tag("Area")
        @DisplayName("Test Rectangle Area")
        void testRectangleArea() {
            var rectangle = setupRectangle();
            double result = rectangle.area();
            double expected = 756.437;
            assertTrue(Math.abs(result - expected) < DELTA);
        }

        @Test
        @Tag("Perimeter")
        @DisplayName("Test Rectangle Perimeter")
        void testRectanglePerimeter() {
            var rectangle = setupRectangle();
            double result = rectangle.perimeter();
            double expected = 112.0;
            assertTrue(Math.abs(result - expected) < DELTA);
        }
    }

    @Nested
    @Tag("Square")
    @DisplayName("Square Tests")
    class SquareTest {
        @Test
        @Tag("Area")
        @DisplayName("Test Square Area")
        void testSquareArea() {
            var square = setupSquare();
            double result = square.area();
            double expected = 105.062;
            assertTrue(Math.abs(result - expected) < DELTA);
        }

        @Test
        @Tag("Perimeter")
        @DisplayName("Test Square Perimeter")
        void testSquarePerimeter() {
            var square = setupSquare();
            double result = square.perimeter();
            double expected = 41.0;
            assertTrue(Math.abs(result - expected) < DELTA);
        }
    }
}

File: elixir/test/shapes/models_test.exs
alias Fixtures.Helper
alias Shapes.Base.Models.{Circle, Rectangle, Square}

defmodule TestsCircle do
  use ExUnit.Case, async: true

  @delta 0.001

  describe "Circle" do
    @tag :area
    test "calculates area" do
      circle = Helper.setup_circle()
      result = Circle.area(circle)
      expected = 754.767
      assert abs(result - expected) < @delta
    end

    @tag :perimeter
    test "calculates circumference" do
      circle = Helper.setup_circle()
      result = Circle.perimeter(circle)
      expected = 97.389
      assert abs(result - expected) < @delta
    end
  end
end

defmodule TestsRectangle do
  use ExUnit.Case, async: true

  @delta 0.001

  describe "Rectangle" do
    @tag :area
    test "calculates area" do
      rectangle = Helper.setup_rectangle()
      result = Rectangle.area(rectangle)
      expected = 756.437
      assert abs(result - expected) < @delta
    end

    @tag :perimeter
    test "calculates perimeter" do
      rectangle = Helper.setup_rectangle()
      result = Rectangle.perimeter(rectangle)
      expected = 112.0
      assert abs(result - expected) < @delta
    end
  end
end

defmodule TestsSquare do
  use ExUnit.Case, async: true

  @delta 0.001

  describe "Square" do
    @tag :area
    test "calculates area" do
      square = Helper.setup_square()
      result = Square.area(square)
      expected = 105.062
      assert abs(result - expected) < @delta
    end

    @tag :perimeter
    test "calculates perimeter" do
      square = Helper.setup_square()
      result = Square.perimeter(square)
      expected = 41.0
      assert abs(result - expected) < @delta
    end
  end
end
